src/main/java/com/qunite/api/service/QueueServiceImpl.java

```diff
@@ -121,6 +122,48 @@ public void deleteById(Long queueId, String principalName) {
     }
   }
 
+  @Override
+  @Transactional
+  public Optional<List<User>> getManagers(Long queueId) {
+    return queueRepository.findById(queueId).map(Queue::getManagers).map(List::copyOf);
+  }
+
+  @Override
+  @Transactional
+  public void addManager(Long managerId, Long queueId, String principalName) {
+    checkIfUserIsAbleToModifyQueueManagers(managerId, queueId, principalName);
+
+    var queue = queueRepository.getReferenceById(queueId);
+    var manager = userRepository.getReferenceById(managerId);
+    if (manager.getUsername().equals(principalName)) {
+      throw new ForbiddenAccessException("You do not need to specify yourself as a manager ;)");
+    }
+    queue.addManager(manager);
+  }
```

Comment:
You are using two `getReferenceById()` methods. This method supposed to be call to lazy load an entity for relationship usage, it means there is no point to lazily load, let's say, queue, when you then call `queue.getCreator()` in `checkIfUserIsAbleToModifyQueueManagers()` method (Thus you are not using repo method as intended).

Problem is two of yours lazily fetched entites are than user eagerly by calling their properties values (in case of manager you call `manager.getUsername()`).
In this scenarion, you have the following Hibernate calls:

![изображение](https://github.com/q-unite/qunite-api/assets/92949251/ce574cab-93e6-4042-9ba0-84d0188f5284)

If we are really talking about optimization in calls to db, I rewrote you method as:
```
@Transactional
  public void addManager(Long managerId, Long queueId, String principalName) {
    var queue = findById(queueId).orElseThrow(
        () -> new QueueNotFoundException("Could not find queue by id: %d".formatted(queueId)));
    var manager = userRepository.getReferenceById(managerId);
    if (!isUserQueueCreatorByCredentials(queueId, principalName)) {
      throw new ForbiddenAccessException("You can not modify this queue");
    }
    if (queue.getCreator().getId().equals(managerId)) {
      throw new ForbiddenAccessException("You do not need to specify yourself as a manager ;)");
    }
    queue.addManager(manager);
  }
```
  
  Here you can see `referenceById()` is now used as intended. Yeah, now method looks more complicated, as with this approach it will be pointless to introduce one more private method (as I have asked you before), but now we have the following result:
  
  
![изображение](https://github.com/q-unite/qunite-api/assets/92949251/c90915e2-5dc5-45c7-9c65-521c08b23bea)

One more thing to note, getReferenceById() throws EntityNoFoundException in case manager does not exist:

![изображение](https://github.com/q-unite/qunite-api/assets/92949251/54021320-61a8-47e1-bc78-63f008603c05)

Therefore, if you stick to use this approach, you are supposed to introduce extra `@ExceptionHandler`. Of course, `deleteManager()` also should be refactored in that case.

Reply:
The least SQL statements could be gained if some code would be duplicated. 
In case we want to avoid code duplication, extra SQL statemets would be executed anyway, because if we want to validate our data, we need to extract it from database, and we can not pass it futher since we have 2 entities(Queue and Manager)(or make private method return something like Entry<Queue, User>?). So we are forced to extract this entities again.
I can not see the solution where we gain the least SQL and without code duplication. It would be grateful if you could suggest one. Unless, there is a question: what do we want more: clean code or 2 less SQL statements?